PerimeterXModule/Internals/PxContext.cs

```diff
@@ -7,167 +7,257 @@
+		public string HttpVersion { get; set; }
+		public string HttpMethod { get; set; }
+		public List<RiskRequestHeader> Headers { get; set; }
+		public string Hostname { get; set; }
+		public string Uri { get; set; }
+		public string UserAgent { get; set; }
+		public string FullUrl { get; set; }
+		public string S2SCallReason { get; set; }
+		public int Score { get; set; }
+		public string Vid { get; set; }
+		public string UUID { get; set; }
+		public BlockReasonEnum BlockReason { get; set; }
+		public bool MadeS2SCallReason { get; set; }
+		public string S2SHttpErrorMessage { get; set; }
+		public string BlockAction { get; set; }
+		public string BlockData { get; set; }
+		public HttpContext ApplicationContext { get; private set; }
+		public bool SensitiveRoute { get; set; }
+		public PassReasonEnum PassReason { get; set; }
+		public long RiskRoundtripTime { get; set; }
+		public CookieOrigin CookieOrigin { get; set; }
+		public Dictionary<string,string> OriginalTokens { get; set; }
+		public IPxCookie OriginalToken { get; set; }
+		public string OriginalTokenError { get; set; }
+		public string OriginalUUID { get; set; }
+		public object DecodedOriginalToken { get; set; }
+
+		public PxContext(HttpContext context, PxModuleConfigurationSection pxConfiguration)
+		{
+			ApplicationContext = context;
+
+			CookieOrigin = CookieOrigin.COOKIE;
+			PxCookies = new Dictionary<string, string>();
+			OriginalTokens = new Dictionary<string, string>();
+			S2SCallReason = "none";
+
+			// Get Headers
+
+			// if userAgentOverride is present override the default user-agent
+			string userAgentOverride = pxConfiguration.UserAgentOverride;
+			if (!string.IsNullOrEmpty(userAgentOverride))
+			{
+				UserAgent = context.Request.Headers[userAgentOverride];
+			}
+			// ua fallback
+			if (string.IsNullOrEmpty(UserAgent))
+			{
+				UserAgent = context.Request.Headers["user-agent"];
+			}
+
+			Headers = new List<RiskRequestHeader>();
+
+			foreach (string header in context.Request.Headers.Keys)
+			{
+				if (!pxConfiguration.SensitiveHeaders.Contains(header))
+				{
+					RiskRequestHeader riskHeader = new RiskRequestHeader
+					{
+						Name = header,
+						Value = context.Request.Headers.Get(header)
+					};
+					if (header.ToLower() == "user-agent")
+					{
+						riskHeader.Value = UserAgent;
+					}
+					Headers.Add(riskHeader);
+				}
+			}
+
+			// Handle Cookies
+			var contextCookie = context.Request.Cookies;
+			string mobileHeader = context.Request.Headers[PxConstants.MOBILE_HEADER];
+			// Check if X-PX-AUTHORIZATION exist
+			if (mobileHeader != null)
+			{
+				// Extract Original Tokens
+				CookieOrigin = CookieOrigin.HEADER;
+				string originalToken = context.Request.Headers[PxConstants.ORIGINAL_TOKEN];
+				if (!string.IsNullOrEmpty(originalToken))
+				{
+					string[] splitedOriginalToken = originalToken.Split(MOBILE_DELIMITER, 2);
+					string[] fallbackSplittedOriginalToken = originalToken.Split(MOBILE_DELIMITER);
```

Comment:
I don't think that you should calculate `fallbackSplittedOriginalToken` "in any case". 
You need it only if the next `if` is `false`.
Instead of starting with a series of `else if`s, have one general `else`, place the `fallbackSplittedOriginalToken` calculation inside, and then follow with specific `if`s that are related only to the resulting `fallbackSplittedOriginalToken`.

Reply:
I'll explain why